### DISCUSSION TITLE
How do we use/support our own cross compilers with vscode

### DISCUSSION THREAD

**Discussion by ganiger143** [microsoft/vscode-cpptools#11992] [General]
We have our own cross compilers and wanted to be recognized by the microsoft C/C++ toolchain for compiling c/c++ programs. How do we do that?

**Comment by Colengms** [comment]
Hi @ganiger143 .  This is the repo for the C/C++ language service extension for VS Code.

If you're referring to (editor) language services, such as IntelliSense, the C/C++ Extension is able to interrogate most clang and gcc based compilers, by passing standard arguments to solicit a list of system includes and predefined macros.  If your compilers are clang or gcc based, it's possible they may be supported when specified in a `compile_commands.json` command line, the `compilerPath` field of a `c_cpp_properties.json` file, or `C_Cpp.default.compilerPath` setting.  (And may require additional args in a corresponding `compilerArgs` field, to configure the compiler's output as desired.)  We do have an open issue https://github.com/microsoft/vscode-cpptools/issues/6931 tracking implementing a mechanism for interrogating compilers that can be extended by end-users to enable support for custom compilers.

If you're referring to use with a build system, that is not something that VS Code or the C/C++ extension provides directly.  There are several extensions such as CMake Tools, Makefile Tools, which provide build system integration (and can also be used to provide IntelliSense configurations to the C/C++ Extension).

**Reply by ganiger143** [reply]
We are not using the clang and gcc compilers, but we have our own cross compilers for C/C++, and we need both editor and build features. What I understand is from your reply is that we cannot extend the IntelliSense, the C/C++ Extension for compilers than the clang and gcc. Is that my understanding is correct?

**Reply by bobbrow** [reply]
> What I understand is from your reply is that we cannot extend the IntelliSense, the C/C++ Extension for compilers than the clang and gcc. Is that my understanding is correct?

Correct. Not until #6931 is complete. Until that time, you would have to list out your system includes and defines explicitly in the "includePath" and "defines" properties of your configurations in `c_cpp_properties.json`.

**Comment by dmitttri** [comment]
Hm, if you are using your custom toolchains (compiler, linker, binutils, etc), then I don't get why you need them to be recognized by Microsoft toolchain? 

But if you just want to use VSCODE as an editor (with Intellisense) to develop and build your cross compiled C/C++ code, then you can apply the same approach I am using these days. Install following extensions:

  -  `C/C++` which brings Intellisense for your code (navigation, auto-complete, etc...) 
  - `VSCode Action Buttons Ext`.  which brings ability to define new custom buttons, and shell commands to be executed (make, cmake, etc..)

Once you install `VSCode Action Buttons Ext`, edit your workspace json file, to add a new section which will define new buttons and corresponding shell commands.

```
        "actionButtons": {
            "defaultColor": "white", 
            "loadNpmCommands": false,
            "commands": [
                {
                    "name": "Build",
                    "color": "yellow",
                    "terminalName": "Build Terminal",
                    "ignoreClear": true,
                    "command": "make"
                },
                {
                    "name": "Clean",
                    "color": "yellow",
                    "terminalName": "Build Terminal",
                    "ignoreClear": true,
                    "command": "make clean",
                }
            ]
        }
```
